openmc/material.py

```diff
@@ -560,6 +560,73 @@ def add_element(self, element, percent, percent_type='ao', enrichment=None):
         for nuclide in element.expand(percent, percent_type, enrichment):
             self.add_nuclide(*nuclide)
 
+    def add_material(self, material, percent, percent_type='ao'):
+        """Mix a material to the material
+        Parameters
+        ----------
+        material : openmc.Material
+            Material to add
+        percent : float
+            Atom or weight percent
+        percent_type : {'ao', 'wo'}
+            'ao' for atom percent and 'wo' for weight percent
+        """
+
+        if self._macroscopic is not None:
+            msg = 'Unable to add a Material to Material ID="{0}" as a ' \
+                  'macroscopic data-set has already been added'.format(self.id)
+            raise ValueError(msg)
+
+        if not isinstance(material, openmc.Material):
+            msg = 'Unable to add a Material to Material ID="{0}" with a ' \
+                  'non-Material value "{1}"'.format(self.id, material)
+            raise ValueError(msg)
+
+        elif not isinstance(percent, Real):
+            msg = 'Unable to add a Material to Material ID="{0}" with a ' \
+                  'non-floating point value "{1}"'.format(self.id, percent)
+            raise ValueError(msg)
+
+        elif self.temperature != material.temperature:
+            msg = 'Unable to add a Material to Material ID="{0}" with a ' \
+                  'different temperature "{1}"'.format(self.id, temperature)
+            raise ValueError(msg)
+
+        elif percent_type not in ['ao', 'wo']:
+            msg = 'Unable to add a Material to Material ID="{0}" with a ' \
+                  'percent type "{1}"'.format(self.id, percent_type)
+            raise ValueError(msg)
+
+        self_percent = 1. - percent
+
+        # Get the atom densities in the given material
+        new_nucs = {}
+
+        # Compute the sum of the atom densities
+        amm = material.average_molar_mass
+        density_sum = 0.
+        for nuc, dens in material.get_nuclide_atom_densities().values():
+            # Convert to weight percent, if requested
+            if percent_type == 'wo':
+                if isinstance(nuclide, openmc.Nuclide):
+                    atomic_mass = openmc.data.atomic_mass(nuclide.name)
+                else:
+                    atomic_mass = openmc.data.atomic_mass(nuclide)
+
+                newdens = dens * atomic_mass / amm
+
+            new_nucs[nuc] = (nuc, newdens)
+            density_sum += newdens
+
+        # Add the nuclide to the material
+        for nuc, dens in new_nucs.values():
+            density = dens / density_sum
+            self.add_nuclide(nuc, density*percent, percent_type)
+
+        # Add the s(a,b) tables
+        for sab in material.sab:
+            self.add_s_alpha_beta(sab)
```

Comment:
Again, we'll need a check for an existing S(a,b) table, and the appropriate fraction should be applied.

Reply:
@paulromano Could you take another look at this? I'm trying to figure out how to ensure the correct thermal scattering cross sections are added when mixing materials. I don't see a method for finding out what nuclides are in a particular S(a,b) table and if multiple nuclides are in a particular table like 'c_BeO' or 'c_C6H6' how are the fractions applied? It seems like things get messy quickly when multiple S(a,b) tables are being mixed with materials that might have the same nuclides but not with S(a,b) tables. Do you have any advice on how to proceed with this?

Reply:
Ugh, yeah I see what you mean. It does start to get messy to cover all the possible scenarios. So as far as figuring out what nuclides are relevant for a particular table, check out [this logic](https://github.com/openmc-dev/openmc/blob/de99faefc93fe46a48fb4c5abcb4b88fe0eec174/openmc/data/thermal.py#L718-L728) which almost solves that. I say _almost_ because you would need to know what the appropriate atomic number is and we don't really have that information at hand.

A reasonable approach would be to look at the first applicable nuclide in the materials being mixed and use the relative densities of that nuclide to determine what fraction of the S(a,b) table to apply. Note that the fractions do not need to sum to unity; say you were mixing two materials that were exactly the same, but one of them used an S(a,b) table. In that case, you would just have `fraction=0.5` on the S(a,b) table for the resulting mixed material.

Another reasonable approach would be to just punt on this and throw an exception saying that we don't support mixing materials with S(a,b) tables :laughing: Not sure if that affects your intended use case.

Reply:
Haha, ok since it doesn't affect my use case yet, I'm going to punt and raise a `NotImplementedError` for now.